File: app-miyou/src/main/java/com/miyou/app/domain/dialogue/service/SentenceAssembler.java
package com.miyou.app.domain.dialogue.service;

import java.util.List;

import reactor.core.publisher.Flux;

public class SentenceAssembler {

	public Flux<String> assemble(Flux<String> tokenStream) {
		return tokenStream.bufferUntil(this::isSentenceEnd).filter(list -> !list.isEmpty())
			.map(this::joinTokensToSentence);
	}

	private boolean isSentenceEnd(String token) {
		if (token == null || token.isEmpty()) {
			return false;
		}
		String trimmed = token.trim();
		return trimmed.endsWith(".") || trimmed.endsWith("!") || trimmed.endsWith("?")
			|| trimmed.endsWith("다.");
	}

	private String joinTokensToSentence(List<String> tokens) {
		return String.join("", tokens).trim();
	}
}

File: domain/src/main/kotlin/com/miyou/app/domain/dialogue/service/SentenceAssembler.kt
package com.miyou.app.domain.dialogue.service

import reactor.core.publisher.Flux

class SentenceAssembler {
    fun assemble(tokenStream: Flux<String>): Flux<String> =
        tokenStream
            .bufferUntil { isSentenceEnd(it) }
            .filter { it.isNotEmpty() }
            .map { joinTokensToSentence(it) }

    private fun isSentenceEnd(token: String): Boolean {
        if (token.isBlank()) {
            return false
        }
        val trimmed = token.trim()
        return trimmed.endsWith(".") || trimmed.endsWith("!") || trimmed.endsWith("?")
    }

    private fun joinTokensToSentence(tokens: List<String>): String = tokens.joinToString(separator = "").trim()
}
